File: cpp/problem874.cpp
#include <mystd.h>
using namespace MYSTD;
using namespace std;

/**
 * just simulate the walk
 * The obstacles are converted to a hash-set for fast looking-up.
 */ 
class Solution {
public:
    int robotSim(vector<int>& commands, vector<vector<int>>& obstacles) {
        std::unordered_set<long> st;
        for(const auto &i: obstacles) st.emplace( long(i[0])*100000 + i[1] );

        int dx = 0, dy = 1, x = 0, y = 0, maxdis = 0;
        for(const auto &command: commands){
            switch(command){
                case -2: swap(dx,dy); dx = -dx; break;
                case -1: swap(dx,dy); dy = -dy; break;
                default:
                    maxdis = max(maxdis, x*x+y*y);
                    for(int i=0; i<command; ++i){
                        if( st.find( long(x+dx)*100000+y+dy ) == st.end() ){
                            x += dx; y += dy;       
                        }else break;
                    }
            }
        }
        maxdis = max(maxdis, x*x+y*y);
        return maxdis;
    }
};

int main(int argc, char const *argv[]){
    Solution sol;

    // cases in Leetcode Problem
    vector<int> commands = {4,-1,3};
    vector<vector<int> > obst;
    cout << sol.robotSim( commands, obst ) << endl;

    commands.assign({4,-1,4,-2,4});
    obst.assign({{2,4}});
    cout << sol.robotSim( commands, obst ) << endl;

    return 0;
}

File: swift/problem874.swift
//swift implementation of cpp version
class Solution {
    func robotSim(_ commands: [Int], _ obstacles: [[Int]]) -> Int {
        //using one set to store both x and y coordinates of obstacles
        let obstacles = obstacles.reduce(into: Set<Int>()){$0.insert($1[0]*100_000+$1[1])}
        
        var x = 0; var y = 0
        var dx = 0; var dy = 1
        var maxDist = 0
        
        for command in commands {
            switch command {
            case -2:
                swap(&dx,&dy); dx = -dx; break
            case -1:
                swap(&dx,&dy); dy = -dy; break
            default:
                maxDist = max(maxDist,x*x+y*y)
                for _ in 0..<command {
                    if !obstacles.contains((x+dx)*100_000+y+dy) {
                        x += dx; y += dy
                    } else {
                        break
                    }
                }
            }
        }
        
        maxDist = max(maxDist,x*x+y*y)
        return maxDist
    }
}
